### ISSUE TITLE
Tutorial: step 3 (Installing as package) needs rationale

### ISSUE THREAD

**Issue by uvtc** [pallets/flask#2421]
Step 3 of the tutorial ([installing flaskr as a package](http://flask.pocoo.org/docs/0.12/tutorial/packaging/)) needs rationale at the top. As it stands, that section of the tutorial tells the user out of the blue that they should install their webapp as a package, but doesn't say why. The user can already run their webapp via `flask run`, so it's not clear why they should at this point set up and install their package into their venv.

**Comment by davidism** [comment]
* It's installed like all other packages, rather than relying on injecting the current directory into `sys.path`. This means that it's importable as long as you're in the virtualenv, rather than needing to be at the root project directory.
* `FLASK_APP` can refer to it by name instead of path (or path to `__init__.py`). (This is true for non-installed packages in the next release, but installing is still recommended.)
* You can install entry points, which is especially helpful if you need to customize the CLI.
* You can deploy like any other Python library, by creating a wheel and installing it. The dev environment now mirrors the prod environment. This also helps you separate your app behavior from your repository structure.
*  In Python 2 especially, it avoids common "everything at the top level" relative import problems.
* You can isolate your tests using tox.

Basically, you get all the benefits of installing packages in general.

**Comment by uvtc** [comment]
Thanks, David. If I can figure out how it works without installing (posting in another thread), I'll create a PR for tutorial.
-- John



On Tue, Jul 18, 2017, at 10:58 AM, David Lord wrote:
>  * It's installed like all other packages, rather than relying on injecting the current directory into sys.path. This means that it's importable as long as you're in the virtualenv, rather than needing to be at the root project directory.
>  * FLASK_APP can refer to it by name instead of path (or path to __init__.py). (This is true for non-installed packages in the next release, but installing is still recommended.)
>  * You can install entry points, which is especially helpful if you need to customize the CLI.
>  * You can deploy like any other Python library, by creating a wheel and installing it. The dev environment now mirrors the prod environment. This also helps you separate your app behavior from your repository structure.
>  * In Python 2 especially, it avoids common "everything at the top level" relative import problems.
>  * You can isolate your tests using tox.> Basically, you get all the benefits of installing packages in general.> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub[1], or mute the thread[2]. > 


Links:

  1. https://github.com/pallets/flask/issues/2421#issuecomment-316091158

**Comment by davidism** [comment]
I don't want the tutorial to not install the project. We want to encourage installation.

**Comment by uvtc** [comment]
This is what I'm trying to understand (just posted to the ML). Isn't it customary to run from the project directory during development, without installing it? Trying to understand how development workflow goes, as opposed to deployment.

My *guess* is that there should be a paragraph at the top of this section saying something about how, in the end, you're going to want to be able to install as a package for a number of reasons, and so should set up your project that way from the start, but then describe how regular devolpment works in the project dir.

Haven't gotten this working, nor finished reading the tut, so will get back to this! :)

**Comment by davidism** [comment]
"Regular development" is installing the package. Yes, we should explain why, but no, we should not say that not installing the package is the way to do things.

**Comment by ThiefMaster** [comment]
During develop you should install it, but with `-e` so the installed package just references your source directory.

**Comment by uvtc** [comment]
BTW, my original comment on this issues was mistaken: as David clarified, I can't run my app with a simple `flask run` without first installing it. I was confused because the Quickstart doc did that with a simple standalone hello.py file, rather than using a package.

**Comment by davidism** [comment]
continued in #2513
